### PR TITLE
Feature/add mz egress ips cidr support

### PR DESCRIPTION
**Pull request by jubrad**
### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

**Comment by def-** [comment]
You probably want to rebase this, the main state it's based on looks really old.

File: misc/kind/cluster.yaml [hunk 1]
```diff
@@ -22,7 +22,7 @@ kubeadmConfigPatches:
         "service-node-port-range": "32000-32063"
 nodes:
   - role: control-plane
-    image: kindest/node:v1.29.7
+    image: kindest/node:v1.31.0
```
Comment (jubrad) [inline misc/kind/cluster.yaml:25]: I can revert this, but something on my laptop (almost certainly my kubectl version) prevented this from working.

File: doc/user/content/sql/system-catalog/mz_catalog.md [hunk 2]
```diff
@@ -189,9 +189,11 @@ The `mz_egress_ips` table contains a row for each potential IP address that the
 system may connect to external systems from.
 
 <!-- RELATION_SPEC mz_catalog.mz_egress_ips -->
-Field       | Type     | Meaning
-------------|----------|--------
-`egress_ip` | [`text`] | The IP address.
+Field           | Type     | Meaning
+----------------|----------|--------
+`egress_ip`     | [`text`] | The start of the range of IP addresses.
+`prefix_length` | [`uint8`]   | The number of leading bits in the CIDR netmask.
```
Comment (benesch) [inline doc/user/content/sql/system-catalog/mz_catalog.md:195]:
```suggestion
`prefix_length` | [`integer`] | The number of leading bits in the CIDR netmask.
```

Unsigned integers aren't well understood by SQL drivers; I think you're better off here using `integer` even though it's not quite as precise as a type.

File: src/adapter/Cargo.toml [hunk 3]
```diff
@@ -15,6 +15,7 @@ arrow = { version = "51.0.0", default-features = false }
 async-trait = "0.1.68"
 bytes = "1.3.0"
 bytesize = "1.1.0"
+cidr = "0.2.3"
```
Comment (benesch) [inline src/adapter/Cargo.toml:18]: Would prefer to use the [`ipnet`](https://docs.rs/ipnet/2.10.0/ipnet/) crate instead. It's the gold standard in the Rust ecosystem for CIDR types AFAIK, and is already in our tree in this repo and in the cloud repo. It should offer all the APIs you're using from the `cidr` crate, plus many more.

File: src/environmentd/src/bin/environmentd/main.rs [hunk 4]
```diff
@@ -194,15 +195,16 @@ pub struct Args {
     /// Wildcards in other positions (e.g., "https://*.foo.com" or "https://foo.*.com") have no effect.
     #[structopt(long, env = "CORS_ALLOWED_ORIGIN")]
     cors_allowed_origin: Vec<HeaderValue>,
-    /// Public IP addresses which the cloud environment has configured for
+    /// Public IP or CIDR which the cloud environment has configured for
```
Comment (benesch) [inline src/environmentd/src/bin/environmentd/main.rs:198]: It's always a CIDR now, right? It's just that you can specify a /32?

```suggestion
    /// Public CIDR which the cloud environment has configured for
```

File: doc/user/content/sql/system-catalog/mz_catalog.md (hunk 2 above)
Comment (benesch) [reply doc/user/content/sql/system-catalog/mz_catalog.md:195]: (Also if you're not already aware, `uint8` is a 64-bit integer, not an 8-byte integer.)

File: src/environmentd/src/bin/environmentd/main.rs (hunk 4 above)
Comment (jubrad) [reply src/environmentd/src/bin/environmentd/main.rs:198]: This actually still allows standard IPs and will just convert them to a CIDR of `/32` this worked out of the box with the `cidr` crate so I went with it. If that's not the case with `ipnet` there's no reason we shouldn't only support cidr ranges.

Comment (benesch) [reply src/environmentd/src/bin/environmentd/main.rs:198]: Ah, gotcha!

File: src/adapter/src/catalog/state.rs [hunk 5]
```diff
@@ -9,10 +9,10 @@
 
 //! In-memory metadata storage for the coordinator.
 
+use ipnet::IpNet;
```
Comment (jkosh44) [inline src/adapter/src/catalog/state.rs:12]: nit: I think this belongs in the next group of imports.

File: src/catalog/src/config.rs [hunk 6]
```diff
@@ -7,8 +7,8 @@
 // the Business Source License, use of this software will be governed
 // by the Apache License, Version 2.0.
 
+use ipnet::IpNet;
```
Comment (jkosh44) [inline src/catalog/src/config.rs:10]: nit: I think this belongs in the next group of imports.

File: src/environmentd/src/bin/environmentd/main.rs [hunk 7]
```diff
@@ -12,8 +12,9 @@
 //! It listens for SQL connections on port 6875 (MTRL) and for HTTP connections
 //! on port 6876.
 
+use ipnet::IpNet;
```
Comment (jkosh44) [inline src/environmentd/src/bin/environmentd/main.rs:15]: nit: I think this belongs in the next group of imports.

File: src/environmentd/src/lib.rs [hunk 8]
```diff
@@ -13,6 +13,7 @@
 //! [differential dataflow]: ../differential_dataflow/index.html
 //! [timely dataflow]: ../timely/index.html
 
+use ipnet::IpNet;
```
Comment (jkosh44) [inline src/environmentd/src/lib.rs:16]: nit: I think this belongs in the next group of imports.

**Review by jkosh44 — APPROVED** [review]
LGTM

**Pull request merged**